orderbook/src/maintenance.rs

```diff
@@ -0,0 +1,37 @@
+use crate::database::Database;
+use crate::event_updater::EventUpdater;
+use crate::orderbook::Orderbook;
+use anyhow::Result;
+use futures::future::join_all;
+use shared::current_block::Maintaining;
+use std::sync::Arc;
+
+/// Collects all service components requiring maintenance on each new block
+pub struct ServiceMaintenance {
+    storage: Arc<Orderbook>,
+    database: Database,
+    event_updater: EventUpdater,
+}
+
+impl ServiceMaintenance {
+    pub fn new(storage: Arc<Orderbook>, database: Database, event_updater: EventUpdater) -> Self {
+        ServiceMaintenance {
+            storage,
+            database,
+            event_updater,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Maintaining for ServiceMaintenance {
+    async fn run_maintenance(&self) -> Result<()> {
+        join_all(vec![
+            self.storage.run_maintenance(),
+            self.event_updater.run_maintenance(),
+            self.database.run_maintenance(),
+        ])
+        .await;
```

Comment:
Still need to `tracing::error!` log the errors.

Reply:
Also, this doesn't correctly propagate any errors.

Reply:
I think if we log them it's fine to not propagate because in that sense the method is always successful. Like before the global maintenance task never returns, it just logs errors and keeps going.

Reply:
I'm almost there... just left with an "unused map"

Reply:
> fine to not propagate because in that sense the method is always successful

Great point. I think in that case we should probably remove the error trace from `CurrentBlackStream::run_maintenance_on_new_block()`.

> just left with an "unused map"

You should probably use a `for ... in` loop here for a couple reasons:
- `map` is lazy, so you actually will probably not run the closure at all here
- `map` is for converting one type to another, not really for executing some code on each element in the iterator. There is [`inspect`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.inspect) which is a little more appropriate for this kind of thing, but still not quite right here because it is also lazy (so you would have to iterate it somehow, either with `collect` or `for ... in`).

I think the easiest approach is really a for loop here:
```rust
for join_all(vec![
        self.storage.run_maintenance(),
        self.event_updater.run_maintenance(),
        self.database.run_maintenance(),
    ])
    .await
{
    if let Err(err) = result {
        tracing::error!("maintenance failed with: {}", err);
    }
}
```

Reply:
So I have made the suggested change here, but my only worry now is the unwrap I had to use in `CurrentBlackStream::run_maintenance_on_new_block`. Is this ok? Should it contain some comment that the errors are handled inside maintainer?

Reply:
I don't think we should `unwrap`. Especially if we plan on using this with other `Maintaining` implementations (like the Balancer Vault event indexer in the solver).

Maybe my comment:

> I think in that case we should probably remove the error trace from CurrentBlackStream::run_maintenance_on_new_block().

Was wrong.

Reply:
> However I can imagine we might not always return Ok in other places.

Yeah, I think that is the catch. In the solver for example, we might just run maintenance on the Vault event updater, which in theory could error.

Reply:
One other approach would be to aggregate all errors into one so it gets logged by the caller.